src/signatures.md

```diff
@@ -264,3 +264,72 @@ it should take `self` by value. Examples:
 
 * Closing a file and returning a result code.
 * A builder-pattern object which spits out the thing it was building. ([Example](https://docs.rs/bindgen/0.59.0/bindgen/struct.Builder.html#method.generate)).
+
+## How do I take a thing, and a reference to something within that thing?
+
+For example, suppose you want to give all of your dogs to your friend, yet also
+tell your friend which one of the dogs is the Best Boy or Girl.
+
+A common use case is if you find yourself destructuring a complex data
+structure to work out what to do with it, then you want to pass the whole thing
+into another function but you don't want to have to repeat all that
+destructuring work.
+
+Generally this is an indication that your types are not split down in the correct
+way:
+
+> This is a decomposition problem. Once you’ve found the correct decomposition, everything
```

Comment:
This doesn't give me any idea what the correct decomposition is, though. Can we make a code sample for this one?